witness_beacon.go

```diff
@@ -101,16 +101,21 @@ func (p *preimageBeacon) LookupPreimage(payHash []byte) ([]byte, bool) {
 	return preimage, true
 }
 
-// AddPreImage adds a newly discovered preimage to the global cache, and also
-// signals any subscribers of the newly discovered witness.
-func (p *preimageBeacon) AddPreimage(pre []byte) error {
+// AddPreimages adds a batch of newly discovered preimages to the global cache,
+// and also signals any subscribers of the newly discovered witness.
+//
+// NOTE: The backing slice of MUST NOT be modified, otherwise the subscribers
+// may be notified of the incorrect preimages.
+func (p *preimageBeacon) AddPreimages(preimages ...[]byte) error {
+	for _, preimage := range preimages {
+		srvrLog.Infof("Adding preimage=%x to witness cache", preimage)
+	}
+
 	p.Lock()
 	defer p.Unlock()
 
-	srvrLog.Infof("Adding preimage=%x to witness cache", pre[:])
-
 	// First, we'll add the witness to the decaying witness cache.
-	err := p.wCache.AddWitness(channeldb.Sha256HashWitness, pre)
+	err := p.wCache.AddWitnesses(channeldb.Sha256HashWitness, preimages...)
```

Comment:
Can this be lifted outside of the lock, as it is happening in a db tx already?

Reply:
Since we never remove from the cache, I believe that's true

Reply:
great suggestion!